### PR TITLE
Allow custom compaction windows on builtin tables

### PR DESCRIPTION
**Pull request by umanwizard**
Ignore all commits except the last. This is rebased on top of #16657 which is why the other commits are there.

See commit message for full explanation.

@antiguru as randomly-selected compute team reviewer.

### Motivation

  * This PR adds a known-desirable feature.

Fixes https://github.com/MaterializeInc/materialize/issues/16357
    

### Tips for reviewer

Ignore all commits except the last. This is rebased on top of #16657 which is why the other commits are there.

See commit message for full explanation.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - None

### CODE REVIEW

File: src/environmentd/tests/sql.rs
Comment (jkosh44) [inline src/environmentd/tests/sql.rs:1159]: As discussed in person, It seems like this is implying that when we create a table we set it's `since` to some time 30 days in the past which I think is potentially wrong.

This is existing behavior, but we probably shouldn't encode it in this test. Instead we should use deterministic timestamps to create a table and then fast forward 30 days. Take a look at `test_timestamp_recovery` for an example.

File: src/adapter/src/coord.rs
Comment (jkosh44) [inline src/adapter/src/coord.rs:807]: It's somewhat inefficient to call this in a loop per timestamp. The underlying controller API can receive a single message with multiple IDs and timestamps. Though this is a failure of the read policy API, ideally `initialize_read_policies` would take some struct that contains `GlobalId`s and a timestamp per `GlobalId`.

It's probably fine for now since it looks like this will only ever iterate 2 times, but we probably are going to want to change this API to avoid slowing down start up.

**Review by jkosh44 — APPROVED** [review]
I think the code all looks fine. I'm somewhat concerned that this adds even more complexity to object read policies, which was already fairly complex. I was hoping that when we added non-default compaction windows back to objects we would come up with a design that makes sense holistically across all object. While this change feels very special cased for builtin tables.

File: src/environmentd/Cargo.toml
Comment (teskje) [inline src/environmentd/Cargo.toml:74]: Can we make this a dev dependency?

**Review by teskje — COMMENTED** [review]
I added a couple comments. The logic changes seem fine to me, but I don't think I fully grok the read policy code, so I'll defer to @jkosh44.

File: src/adapter/src/catalog/builtin.rs
Comment (umanwizard) [reply src/adapter/src/catalog/builtin.rs:1595]: yes, I spoke about this with Moritz. He had the same concern as you. I will add an `ALTER SYSTEM` knob for this.

File: src/adapter/src/catalog.rs
Comment (umanwizard) [reply src/adapter/src/catalog.rs:1456]: We did consider this (with @jkosh44); however, we realized that if we open this to users, and a user manually sets 1s (the same value as the default), then if we ever want to change the default, we will not be able to tell whether tables with 1s compaction window should be changed (because they were just inheriting the default) or left with 1s (because a user set them).

So I decided to leave this as optional, to communicate that things with `None` should inherit whatever the default is, including if it's updated to something other than 1s.

Comment (teskje) [reply src/adapter/src/catalog.rs:1456]: Okay, that makes sense!

**Comment by umanwizard** [comment]
The updated version of the code works a bit differently:

1. Instead of hardcoding a duration in the builtin tables, we flag a few of them (and also a builtin source) as being a "retained metrics table".
2. The compaction window of all retained metrics tables is controlled by a system var, so we can modify it with `ALTER SYSTEM ...`
3. The `bootstrap` function no longer needs special logic to fill in non-default values for tables specifically. It just asks the catalog item what its initial compaction window should be, and follows that instruction. Thus, the special logic for tables and sources now lives in `impl CatalogItem`, which is slightly less confusing.

**Comment by umanwizard** [comment]
It should be ready for another look.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [inline src/adapter/src/catalog.rs:1911]: Does this function always return `Some`? If so, should it just have a return type of `Duration`?

File: src/adapter/src/catalog.rs
Comment (jkosh44) [inline src/adapter/src/catalog.rs:1491]: It seems a bit redundant to have both of these because the only two valid combinations are `(Some(_), true)` or `(None, false)`.

File: src/adapter/src/coord.rs
Comment (jkosh44) [inline src/adapter/src/coord.rs:163]: We're trying to avoid using `as`. Is this a `u128` to `u64` conversion? If so I think a `try_from`/`try_into` followed by an `expect` would be better.

File: src/adapter/src/coord.rs
Comment (jkosh44) [inline src/adapter/src/coord.rs:525]: Can you replace these `unwrap`s with `expect`s?

File: src/adapter/src/coord.rs
Comment (jkosh44) [inline src/adapter/src/coord.rs:651]: I'm having trouble following this, why do all sources, tables, and indexes have an initial compaction window? I thought only those that were considered retained_metrics_relations had initial compaction windows.

File: src/adapter/src/coord.rs
Comment (jkosh44) [inline src/adapter/src/coord.rs:823]: Can you add a comment about what we discussed for this that says we want to eliminate the loop in the future?

File: src/adapter/src/coord/sequencer.rs
Comment (jkosh44) [inline src/adapter/src/coord/sequencer.rs:3665]: We should convert `update_storage_base_read_policy` to take a `Vec` of IDs and policies and only send one message to storage instead of sending one message per id.

File: src/adapter/src/catalog.rs
Comment (jkosh44) [reply src/adapter/src/catalog.rs:1911]: Nevermind, I see right there that it can return `None`

File: src/adapter/src/catalog.rs
Comment (umanwizard) [reply src/adapter/src/catalog.rs:1491]: That's true, but we might someday expose the ability to set logical compaction windows to users, but the corresponding objects would still have `is_retained_metrics_relation: false`

File: src/adapter/src/coord.rs
Comment (umanwizard) [reply src/adapter/src/coord.rs:163]: See discussion [here](https://materializeinc.slack.com/archives/CMH6PG4CW/p1671480553139239).

It's a u128 -> u64, but I can't use `try_from` because that's not a const function.

Once Petros lands some changes to `ore`, I'll be able to do this without `as`.

File: src/adapter/src/coord.rs
Comment (umanwizard) [reply src/adapter/src/coord.rs:651]: `initial_logical_compaction_window` always returns `Some` for those types.

If `custom_logical_compaction_window` is not filled in, it just returns the default. It should only return `None` for objects that don't have a notion of compaction window at all (e.g. Views)

File: src/adapter/src/coord.rs
Comment (umanwizard) [reply src/adapter/src/coord.rs]: Actually, there's a way to do it without `as` now, if I just factor out the integer millis value (see latest commit)

**Review by jkosh44 — APPROVED** [review]
LGTM

**Pull request merged**
